### PR TITLE
boards: Add BBC MicroBit V2 support

### PR DESCRIPTION
**Pull request by LingaoM**
The BBC micro:bit v2 is a mini-computer that has been
designed to make the coding fun and easy to learn.

The micro:bit v2 is completely programmable so you can
easily bring your ideas to life! From making games to
creating music and even controlling robots.

The micro:bit comes with neat hardware such as a 25 LED
display, buttons, in-built speakers, Bluetooth 5 & Mesh
connectivity and sensors for temperature, motion & light.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
Thanks for this.  Please review the devicetree pin assignments, and change the compatible to match the vendor name more closely.

**Review by pabigot — CHANGES_REQUESTED** [review]

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]

File: drivers/display/mb2_display.c
Comment (LingaoM) [reply drivers/display/mb2_display.c:1]: In fact, there are still some differences, except what you said, but it’s better to separate the common parts.

Comment (LingaoM) [reply drivers/display/mb2_display.c:1]: @vanwinkeljan  Since the first generation is a 3x9 LED array, which is all on the same GPIO port, while the second generation uses a 5x5 LED array, which is not on the same GPIO port, I have changed the structure `struct mb_display`, so it is not recommended to split it.

File: boards/arm/bbc_microbit2/bbc_microbit2.dts
Comment (pabigot) [inline boards/arm/bbc_microbit2/bbc_microbit2.dts:58]: No space in the compatible string.

**Review by pabigot — CHANGES_REQUESTED** [review]
I tend to agree on refactoring the display code to be shared.  I also don't think it should be necessary to split the pin masks into two `uint32_t` values, one for each port.  Only one pin is on port1, and the corresponding pin on port0 is not used for the display, so just putting all pins into one word, then using a mask to filter them per-port would be more flexible.  Also there's a lot of extra noise defining names for the LED pins when those names are used only to initialize a static array of pin numbers.  (A static array of pin masks would probably reduce code size and increase speed.)

However, I haven't dived into the code enough to figure out what it's doing to confirm that would work.  The whole display driver could do with an overhaul and tune up, though that's a pretty big change affecting both microbit platforms.  So I'm not making that one of my must-change things.

The compatible has to be fixed, though.

**Comment by LingaoM** [comment]
> I tend to agree on refactoring the display code to be shared. I also don't think it should be necessary to split the pin masks into two `uint32_t` values, one for each port. Only one pin is on port1, and the corresponding pin on port0 is not used for the display, so just putting all pins into one word, then using a mask to filter them per-port would be more flexible. Also there's a lot of extra noise defining names for the LED pins when those names are used only to initialize a static array of pin numbers. (A static array of pin masks would probably reduce code size and increase speed.)

There is something worthy of optimization, but I still think the existing logic structure is easier to read. If you can provide a better way, you are welcome to make further changes.

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]
i will not accept the display driver changes as is

**Comment by LingaoM** [comment]
> i will not accept the display driver changes as is

Okay, I will try my best to disassemble the common parts.

File: drivers/display/mb2_display.c
Comment (jfischer-no) [reply drivers/display/mb2_display.c:1]: I suggest to remove mb display relevant parts and make it into a separate PR, which would introduce a generic LED matrix driver or something like that.

Comment (LingaoM) [reply drivers/display/mb2_display.c:1]: Okay

File: boards/arm/bbc_microbit2/doc/index.rst
Comment (jfischer-no) [inline boards/arm/bbc_microbit2/doc/index.rst:61]: Please remove Display from this table as well.

Comment (pabigot) [reply boards/arm/bbc_microbit2/doc/index.rst:61]: I concur; this is the last piece I see missing before I could approve this.

**Comment by tagunil** [comment]
Shouldn't the board directory and the dts file also be renamed to bbc_microbit_v2 as in BOARD Kconfig?

**Comment by henrikbrixandersen** [comment]
This sounds like a candidate for the newly added "Multiple board revisions" support:
https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#multiple-board-revisions

**Comment by LingaoM** [comment]
When I was looking at this revision, I think the difference between the two boards is so great that there is no need to adopt a revision method. The chips on these two boards have changed!

I think this method is more appropriate if only some simple external definitions are changed, such as gpio pins!

**Comment by LingaoM** [comment]
The revision method distinguishes different boards by specifying different overlay files, but this file is only suitable for some simple peripheral changes! https://docs.zephyrproject.org/latest/guides/dts/howtos.html#use-devicetree-overlays

**Review by henrikbrixandersen — APPROVED** [review]

**Review by vanwinkeljan — APPROVED** [review]

File: boards/arm/bbc_microbit2/bbc_microbit2.dts
Comment (LingaoM) [reply boards/arm/bbc_microbit2/bbc_microbit2.dts]: Done.

**Review by carlescufi — CHANGES_REQUESTED** [review]
This is called the "new BBC micro:bit", so I must insist that we use the board revision feature here, even though the hardware layout is completely different.
https://microbit.org/new-microbit/

**Review by pabigot — APPROVED** [review]

**Comment by pabigot** [comment]
I don't strongly object to delaying this until it's reworked to support the new board variant infrastructure, but I'm not convinced that the design of that infrastructure matches well with what Zephyr needs.

Microbit has at least [four documented board revisions](https://tech.microbit.org/hardware/i2c-shared/): 1.3, 1.5v1, 1.5v2, 2.  By those numbers there must be several other board revisions.

If using the new infrastructure (which seems to want three-element version identifiers) means creating four board revisions to cover these cases, it's not obvious how to design a board revision that supports all rev 1 boards, which is certainly technically possible and is highly desirable, as it matches the use case of application firmware intended to support multiple fielded product revisions.

Imposing this requirement seems a pretty high new bar.  I don't see any in-tree boards that use it yet that could serve as examples, and we don't seem to have a plan to update the existing boards to make use of this feature, which would be a good step toward validating the solution.

**Comment by carlescufi** [comment]
> I don't strongly object to delaying this until it's reworked to support the new board variant infrastructure, but I'm not convinced that the design of that infrastructure matches well with what Zephyr needs.
> 
> Microbit has at least [four documented board revisions](https://tech.microbit.org/hardware/i2c-shared/): 1.3, 1.5v1, 1.5v2, 2. By those numbers there must be several other board revisions.
> 
> If using the new infrastructure (which seems to want three-element version identifiers) means creating four board revisions to cover these cases, it's not obvious how to design a board revision that supports all rev 1 boards, which is certainly technically possible and is highly desirable, as it matches the use case of application firmware intended to support multiple fielded product revisions.
> 
> Imposing this requirement seems a pretty high new bar. I don't see any in-tree boards that use it yet that could serve as examples, and we don't seem to have a plan to update the existing boards to make use of this feature, which would be a good step toward validating the solution.

We don't currently support the 1.x variants, but that is something that the board revision feature supports natively. Once we implement it for this board, the user will potentially be able to do this:

```
west build -b bbc_microbit@1  # whatever 1.x revision we deem to be the default
west build -b bbc_microbit@2 # The "new micro:bit" as they describe in the webpage, i.e. the one added by this PR
west build -b bbc_microbit@1.3 # If we later add support for 1.x variants
```

The reason I'd like to use board revisions for this particular board is that its name has not changed. It's still the BBC micro:bit, they just call it "new". This means that having a new board is counterintuitive from a naming perspective.

**Comment by henrikbrixandersen** [comment]
How do we proceed to get this resolved?

**Comment by pabigot** [comment]
> How do we proceed to get this resolved?

IMO: Merge this, then develop a plan to go through the existing boards during 2.6 development to determine which of them should be converted to support multiple versions using the new API, and do so widely at that time.

I wasn't able to find any uses of this feature in-tree.  I really think it's a little too much to insist a new contribution be the first to adapt to new infrastructure that hasn't been validated on existing boards, such as the three Particle feather variants, especially when it requires extra work to change an existing board.

**Comment by pabigot** [comment]
> The reason I'd like to use board revisions for this particular board is that its name has not changed. It's still the BBC micro:bit, they just call it "new". This means that having a new board is counterintuitive from a naming perspective.

BBC [do describe it](https://tech.microbit.org/latest-revision/announcement/) as "micro:bit V2" in contexts where the differences become important.

I agree that due to their target users  "micro:bit" will remain the generic name, but the version is very important for its functionality.  While other development systems like Arduino might be able to figure out the variant automatically that will not be the case with Zephyr.  It must be expressed in the board identification, whether that's `bbc_microbitv2` or `bbc_microbit@2`.

**Comment by carlescufi** [comment]
> IMO: Merge this, then develop a plan to go through the existing boards during 2.6 development to determine which of them should be converted to support multiple versions using the new API, and do so widely at that time.

OK, I think this is a reasonable approach, especially given that I thought I would have the time to finish my in-progress port using board versions but reality has shown me repeatedly that something keeps coming up preventing me from finishing this work. I will merge this and then, if I finally get the time, send a PR to transform this to use board versions.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
